### ISSUE TITLE
Gradient of SVD with degenerate singular values becomes NaN

### ISSUE THREAD

**Issue by Jakob-Unfried** [google/jax#2311]
If an SVD has degenerate singular values (multiple entries of S are exactly equal), the gradient pass will give `nan`.

This is because the [AD formula](https://arxiv.org/pdf/1909.02659.pdf) contains something like
`1 / (s[i] ** 2 - s[j] ** 2) ` which becomes `nan` if `s[i] == s[j]`

In particular, this situation arises when multiple singular values are `0.0`. and it evaluates the entire gradient of a function to `nan`, even if the function does not depend on the rows of U and columns of V that correspond to the `0.0` singular values (See example below). Essentially in this case, even if the adjoint gradients of the corresponding slices of U and V are `0.`, they are multiplied by `nan` and thus give `nan`.

I don't have a clear vision, what the best case scenario is here. 
I just wanted to share that this problem arises for me and how i will dirty-fix it.
Maybe fix "1." would be of interest to implement in jax? thoughts?

### Example
I contruct a matrix that has only a few non-zero singular values (actually, only the first in non-zero, but due two numerical error, the second is non-zero too)
```python 
mat_degenerate = np.ones([5, 5])
U, S, V = np.linalg.svd(mat_degenerate)
print(S)
>>> [4.9999995e+00 1.9792830e-07 0.0000000e+00 0.0000000e+00 0.0000000e+00]
```

I design a function that only depends on the first two singular values (which are clearly neither degenerate nor zero), and the corresponding rows (columns) of U (V)

```python
def truncated_svd(x, num):
    U, S, V = np.linalg.svd(x, full_matrices=False)
    return U[:, :num], S[:num], V[:num, :]

def foo(x):
    U, S, V = truncated_svd(x, 2)
    return np.real(np.trace(np.dot(V, U))) + np.sum(S)
```

And see that the *entire* gradient becomes `nan`

```python
print(value_and_grad(foo)(mat_degenerate)
>>> (DeviceArray(6.9999995, dtype=float32), DeviceArray([[nan, nan, nan, nan, nan],
              [nan, nan, nan, nan, nan],
              [nan, nan, nan, nan, nan],
              [nan, nan, nan, nan, nan],
              [nan, nan, nan, nan, nan]], dtype=float32))
```

### Dirty workaround
1. for my specific situation:
I will define a custom gradient for `truncated_svd` that avoids the multiplication with `nan`.
Let me know if someone is interested in that.

2. for more general situations, possibly with degenerate but non-zero singular values:
I have seen people use `save_inverse = lambda x, eps: x / (x**2 + eps)` and replace `1 / (s[i] ** 2 - s[j] ** 2) ` by `save_inverse(s[i] ** 2 - s[j] ** 2)`, usually with `eps == 1e-12`

**Comment by Aminpdi** [comment]
Hi Jakob

I have the same problem with backpropagation on SVD.  specifically, when the matrix is large. Did your implementation work? I mean Did you have stable backpropagation and the loss is decreasing properly? If yes, is it possible to share your implementation or explain it?

**Comment by Jakob-Unfried** [comment]
@Aminpdi 

Sending you a PN, I don't want to think about licensing and it being "clean enough", so I don't want to make it public.

**Comment by quantshah** [comment]
Hi, I have a somewhat similar issue with Jax's SVD decomposition for larger matrices. I wrote a simple test to compare against Numpy's SVD and for larger matrices there is a discrepancy. If I check the decomposition and compare, there seems to be errors in the decomposed matrices (the errors depend on the singular values). I have a routine where I am forcing an operator to become positive after SVD by discarding the negative singular values, but this leads to some NaNs later in my code. 

It seems that jax.linalg.svd is tested for matrices of sizes 29 and 53 and with a tolerance of 1e-4, see the [test](https://github.com/google/jax/blob/master/tests/linalg_test.py#L584).

```
import numpy as np
from scipy.stats import unitary_group
from jax import numpy as jnp
np.random.seed(42)

def rand_positive(N):
    """
    Creates a random positive operator.
    
    Args:
        N (int): Size of the operator.
        
    Returns:
        op (ndarray): A positive operator of dimension (N, N)
    """
    U = unitary_group.rvs(N)
    V = unitary_group.rvs(N)
    s = np.zeros((N, N))
    np.fill_diagonal(s, np.random.uniform(-5, 5, size = (N, N)))
    op = np.dot(U@s, np.conjugate(V).T)
    
    return op


def test_svd(N):
    """
    Tests if SVD decomposition from Jax matches with Numpy for a
    random operator of size (N, N).
    
    Args:
        N (int): The size of the operator.
        
    Returns:
        assert_list ([bool, bool, bool]): The assertion for how close the decomposition
                                          O = USV matches with Numpy's decomposition.
    """
    E = rand_positive(N)

    u, s, vh = np.linalg.svd(E)
    ju, js, jvh = jnp.linalg.svd(E)

    return (np.allclose(u, ju, atol=1e-4, rtol=1e-4),
            np.allclose(s, js),
            np.allclose(vh, jvh, atol=1e-4, rtol=1e-4))


test_svd(2), test_svd(3), test_svd(5), test_svd(7), test_svd(29), test_svd(53)
```

which gives output

```
((True, True, True),
 (True, True, True),
 (True, True, True),
 (True, True, True),
 (False, True, False),
 (False, True, False))
```

However, setting the seed to 
```
np.random.seed(29)
```
leads to the failure of `test_svd(53)`.

Is this something that is expected? I am not an expert in the numerics of SVD but it seems that the SVD is incorrect if compared to Numpy's implementation for certain cases. If you think this is an issue, I could open a new thread regarding this.

Meanwhile any help in dealing with this would be very much appreciated. Thanks.

**Comment by hawkinsp** [comment]
@quantshah Please open a new issue for a new problem.

That said, I suspect the issue is you need to enable 64-bit precision in JAX (https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision ). Does that help?

**Comment by quantshah** [comment]
@hawkinsp Thanks for the quick response. I wrote a test using complex128 and enabling 64-bit precision and the errors still remain

```
import numpy as np
from scipy.stats import unitary_group

from jax.config import config
from jax import numpy as jnp

config.update("jax_enable_x64", True)
np.random.seed(29)


def rand_positive(N):
    """
    Creates a random positive operator.
    
    Args:
        N (int): Size of the operator.
        
    Returns:
        op (ndarray): A positive operator of dimension (N, N)
    """
    U = unitary_group.rvs(N)
    V = unitary_group.rvs(N)
    s = np.zeros((N, N))
    np.fill_diagonal(s, np.random.uniform(1., 5., size = (N, N)))
    op = np.dot(np.dot(U, s), np.conjugate(V).T)
    
    return op



N = 20
E = rand_positive(N).astype(jnp.complex128)
print(E.dtype)

u, s, vh = np.linalg.svd(E)
ju, js, jvh = jnp.linalg.svd(E)

np.allclose(u, ju), np.allclose(s, js), np.allclose(vh, jvh)
```

If I actually try to see the difference between the real parts of the matrices then it is quite significant compared to what Numpy gives. I will open a new issue for this if you think it is a bug and requires some further discussions. Thanks.

![Unknown](https://user-images.githubusercontent.com/6968324/108519530-ffc49b80-72c9-11eb-857f-a52b58ea9bd4.png)
![Unknown-1](https://user-images.githubusercontent.com/6968324/108519540-018e5f00-72ca-11eb-8678-a94550033f84.png)

**Comment by souryadey** [comment]
@Jakob-Unfried : I am facing the same problem and like your `safe_inverse` solution, but I'm not sure how to apply such a thing when doing backpropagation in Pytorch. Do you have any idea on that?

**Comment by wangleiphy** [comment]
> @Jakob-Unfried : I am facing the same problem and like your `safe_inverse` solution, but I'm not sure how to apply such a thing when doing backpropagation in Pytorch. Do you have any idea on that?

see here https://github.com/wangleiphy/tensorgrad/blob/master/tensornets/adlib/svd.py

**Comment by wangleiphy** [comment]
Here is an example of how to use that function 
https://github.com/wangleiphy/tensorgrad/blob/master/tensornets/trg.py#L3-L4

**Comment by souryadey** [comment]
That worked, thank you so much!

**Comment by rrunix** [comment]
Hi @Jakob-Unfried @Aminpdi, running into the same issue.. I do not how to apply the "safe_tensor" hack.. Tried porting the tensornets implementation with no success. Could you share this code? Thanks

**Comment by JadM133** [comment]
Hello,
It's been a while since this issue was discussed. But I wanted to share the following paper:
[Robust differentiable SVD](https://arxiv.org/abs/2104.03821).
Which proposes a way of using Taylor's extension to avoid the explosion of gradients when eigenvalues are close (or exactly the same).
We do lose some parts of the gradients, though. Maybe it is possible to use this implementation only when eigenvalues are close?

**Comment by EmilianoG-byte** [comment]
Hi @Jakob-Unfried,

I have sent you a message on your TUM email asking about your solution - I also went to TUM btw :)

**Comment by Alekkk777** [comment]
**Working implementation for complex truncated SVD VJP in JAX**

We ran into this exact issue while training an MPS Physics-Informed 
Neural Network for quantum tunneling (qpinn: github.com/[tuo-user]/qpinn).

The failure mode is physical, not just numerical: during tunneling, 
the wavefunction becomes (|L⟩+|R⟩)/√2, forcing exact degeneracy in 
the Schmidt spectrum (S[0]=S[1]). Standard jax.numpy.linalg.svd VJP 
→ NaN at this point. This is the most physically interesting moment 
of the simulation.

**Fix:** gauge-invariant SVD VJP with Lorentz regularization for 
complex rectangular matrices, extending Liao et al. (2019) 
[Phys. Rev. X 9, 031041] to the complex case.

Three bugs vs naive porting of real→complex:
1. Convention mismatch: JAX cotangents are bilinear (conjugate of 
   Euclidean grad) — requires conjugating dU, dVt before Townsend formula
2. LAPACK phase ambiguity: sign-fixing per column of U required for 
   stable finite-difference checks
3. Lorentz regularization: σᵢ/(σᵢ²-σⱼ²+ε) instead of 1/(σᵢ²-σⱼ²)

**Quantified:** ratio native/qtens = 1.000 for gap > 1e-12, 
diverges only below ~1e-14 (machine precision). Native → NaN at 
exact degeneracy; Qtens → bounded.

Code: github.com/GantryGraph/Qtensx
Real-world validation: https://github.com/GantryGraph/Qpinn
Archived: 10.5281/zenodo.21083309

Happy to open a PR if there's interest.

**Comment by tatopenn-cell** [comment]
Note up front: our fix targets `jnp.linalg.eigh`'s gradient rule specifically, not `svd` directly -- this doesn't close this issue, which is about SVD. Posting in case the same masking approach is a useful starting point for whoever tackles the SVD case, since the failure mode is the same shape (`1/(sigma_i - sigma_j)` for SVD vs `1/(lambda_i - lambda_j)` for eigh, both blowing up at degeneracy).

@shoyer's option 2 above (custom derivative rules for downstream well-defined quantities) is what we ended up shipping.

We needed `jax.grad` through Uhlmann fidelity `F(rho_A, rho_B) = Tr(sqrt(sqrt(rho_A) rho_B sqrt(rho_A)))^2` between density matrices in a quantum-circuit simulator, where `rho_A` is frequently (near-)degenerate in practice -- e.g. the fully mixed state, or any noisy near-pure state with several near-zero, near-degenerate eigenvalues. `jnp.linalg.eigh`'s built-in gradient rule divides by `lambda_i - lambda_j` and returns NaN exactly there.

Fix is the practical masking variant mentioned in Kasim, "Derivatives of partial eigendecomposition of a real symmetric matrix for degenerate cases" (arXiv:2011.04366): a `jax.custom_jvp` on `eigh` that masks the `1/(lambda_i - lambda_j)` term to 0 for near-degenerate pairs instead of letting it blow up. Not Kasim's fuller per-degenerate-block treatment (which recovers a generally nonzero contribution from within the degenerate eigenspace itself for some perturbation directions) -- sufficient here because the only downstream use of the eigenvectors is rebuilding a matrix square root via its trace, so a masked (if not maximally sharp) subgradient is still mathematically valid.

**Edit**: I initially wrote below that this masking approach was "also used by PyTorch's `torch.linalg.eigh` gradient notes and the `xitorch` library's `symeig`" -- that was wrong, I hadn't actually checked, and I should have. Having now read the real source: PyTorch's current `linalg_eig_jvp`/`linalg_eig_backward` (`torch/csrc/autograd/FunctionsManual.cpp`) divides directly by `(L_j - L_i)` with no degenerate-pair masking at all -- same singularity as JAX's default, unresolved. xitorch's own derivation notes (`doc/notes/deriv_symeig.rst`, by the same Kasim) state up front "This derivation assumes the eigenvalues are all unique. Cases with degenerate eigenvalues are treated differently" without giving that treatment on that page, and `symeig`'s own docstring warns directly: "If `s` contains very small numbers or degenerate values, the calculation and its gradient might be inaccurate." So this appears to still be an open problem across the ecosystem, not something already solved elsewhere that we ported -- correcting the record rather than leave an unverified claim standing.

Source + full docstring: https://github.com/tatopenn-cell/Dense-Evolution/blob/1cb93a61b71a185222618601efb220d327e6998a/dense_evolution/mitigation.py#L303-L347

Verified two ways: (1) directional-derivative/JVP comparison against symmetric-tangent finite differences (avoids eigenvector sign/ordering ambiguity) -- exact match, diff=0.000000, on non-degenerate, 2-fold, and 3-fold degenerate test matrices; (2) the fully-mixed-state gradient (worst case: every eigenvalue tied) goes from all-NaN to finite, confirmed directly rather than assumed from the general theory.
